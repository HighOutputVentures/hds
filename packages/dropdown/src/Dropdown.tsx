import {
  chakra,
  Divider,
  Menu,
  MenuButton,
  MenuGroup,
  MenuItem,
  MenuList,
  PlacementWithLogical,
  Portal,
  useDisclosure,
  UseDisclosureReturn,
} from "@chakra-ui/react";
import * as React from "react";
import { v4 as uuid } from "uuid";
import { useStyles } from "./hooks";

interface Item {
  icon?: JSX.Element;
  label: React.ReactNode;
  command?: string;
  onClick?(): void;
}

interface BaseProps {
  placement?: PlacementWithLogical | undefined;
  closeOnSelect?: boolean;
  renderHeader?: JSX.Element;
  renderOption?(item: Item): JSX.Element;
  children(context: UseDisclosureReturn): JSX.Element;
  __menuTestId?: string;
  __menuItemTestId?: string;
}

interface GroupProps {
  items: Item[][];
  isGrouped: true;
}

interface SingleProps {
  items: Item[];
  isGrouped?: false;
}

export type DropdownProps = (GroupProps | SingleProps) & BaseProps;

const Dropdown: React.FC<DropdownProps> = (props) => {
  const {
    children,
    placement,
    renderHeader,
    renderOption = (item) => item.label,
    closeOnSelect,
    __menuTestId = "hds.dropdown",
    __menuItemTestId = "hds.dropdown.item",
    ...others
  } = props;

  const styles = useStyles();
  const disclosure = useDisclosure();

  return (
    <Menu
      variant="unstyled"
      isOpen={disclosure.isOpen}
      onOpen={disclosure.onOpen}
      onClose={disclosure.onClose}
      placement={placement}
      closeOnBlur
      closeOnSelect={closeOnSelect}
      isLazy
      lazyBehavior="keepMounted"
    >
      <MenuButton as={createRefReciever(children(disclosure))} />

      <Portal appendToParentPortal={false}>
        <MenuList sx={styles.menulist} data-testid={__menuTestId}>
          {!!renderHeader && (
            <MenuGroup sx={styles.menugroup}>
              <chakra.div sx={styles.header}>{renderHeader}</chakra.div>
              <Divider sx={styles.divider} />
            </MenuGroup>
          )}

          {!others.isGrouped && (
            <MenuGroup sx={styles.menugroup}>
              {others.items.map((item) => {
                return (
                  <MenuItem
                    sx={styles.menuitem}
                    key={uuid()}
                    onClick={item.onClick}
                    /**
                     *
                     * Only render icon and command if
                     * user did not provide the "renderOption" prop
                     *
                     */
                    {...(!props.renderOption && {
                      icon: item.icon,
                      command: item.command,
                    })}
                    data-testid={__menuItemTestId}
                  >
                    {renderOption(item)}
                  </MenuItem>
                );
              })}
            </MenuGroup>
          )}

          {props.isGrouped &&
            props.items.map((items, idx_0, arr_0) => {
              return (
                <MenuGroup key={uuid()} sx={styles.menugroup}>
                  {items.map((item, idx_1, arr_1) => {
                    const hasDivider =
                      idx_0 < arr_0.length - 1 && idx_1 === arr_1.length - 1;

                    return (
                      <React.Fragment key={uuid()}>
                        <MenuItem
                          icon={item.icon}
                          sx={styles.menuitem}
                          data-testid={__menuItemTestId}
                          /**
                           *
                           * See comment above
                           *
                           */
                          {...(!props.renderOption && {
                            icon: item.icon,
                            command: item.command,
                          })}
                        >
                          {renderOption(item)}
                        </MenuItem>

                        {hasDivider && <Divider sx={styles.divider} />}
                      </React.Fragment>
                    );
                  })}
                </MenuGroup>
              );
            })}
        </MenuList>
      </Portal>
    </Menu>
  );
};

function createRefReciever(children: JSX.Element) {
  return React.forwardRef<any>(function Tunnel(_, ref) {
    return React.cloneElement(children, { ref });
  });
}

export default Dropdown;
