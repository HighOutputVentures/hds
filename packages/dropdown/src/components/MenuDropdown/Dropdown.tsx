import { ChevronDownIcon, ChevronUpIcon } from '@chakra-ui/icons';
import {
  Box,
  Button,
  Divider,
  Menu,
  MenuButton,
  MenuGroup,
  MenuItem,
  MenuList,
  PlacementWithLogical,
  UseDisclosureReturn,
  useDisclosure,
} from '@chakra-ui/react';
import { ButtonVariantsTypes } from '@highoutput/hds/src/theme/components/button/variants';
import React, { FC } from 'react';

export interface Items {
  label: string;
  command?: string;
  icon?: JSX.Element;
}
export interface IMenuItemsSingle {
  items: Array<Items>;
}

export interface IMenuItemsGroup {
  items: Array<Array<Items>>;
}
export interface DropdownFieldMainProps {
  menuButtonText?: string;
  profileUrl?: string;
  indicator?: boolean;
  renderHeader?: React.ReactNode;
  __menuTestId?: string;
  __menuButtonTestId?: string;
  __menuListTestId?: string;
  gap?: string;
  placement?: PlacementWithLogical | undefined;
  variant?: ButtonVariantsTypes;
  closeOnSelect?: boolean;
  showDropdownIndicator?: boolean;
  children(context: UseDisclosureReturn): JSX.Element | null;
}
export interface DropdownFieldPropsGroup extends DropdownFieldMainProps {
  group: true;
  items: IMenuItemsGroup['items'];
  hasDivider?: never;
}
export interface DropdownFieldPropsSingle extends DropdownFieldMainProps {
  group: false;
  items: IMenuItemsSingle['items'];
  hasDivider?: boolean;
}

export type DropdownFieldProps =
  | DropdownFieldPropsGroup
  | DropdownFieldPropsSingle;
const Dropdown: FC<DropdownFieldProps> = (props) => {
  const {
    renderHeader,
    closeOnSelect = true,
    showDropdownIndicator = false,
    items,
    gap,
    group,
    hasDivider,
    children,
    __menuTestId,
    __menuButtonTestId,
    __menuListTestId,
    placement,
    variant = 'unstyled',
  } = props;
  const disclosure = useDisclosure();
  return (
    <Menu
      data-testid={__menuTestId ?? 'hds.menu.dropdown'}
      placement={placement}
      closeOnSelect={closeOnSelect}
    >
      <Box w="full">
        <MenuButton
          variant={variant}
          as={Button}
          isActive={disclosure.isOpen}
          rightIcon={
            !disclosure.isOpen && showDropdownIndicator ? (
              <ChevronDownIcon width="20px" height="20px" />
            ) : disclosure.isOpen && showDropdownIndicator ? (
              <ChevronUpIcon width="20px" height="20px" />
            ) : undefined
          }
          data-testid={__menuButtonTestId}
          display="flex"
        >
          {children(disclosure)}
        </MenuButton>

        <MenuList
          boxShadow={
            '0px 12px 16px -4px rgba(16, 24, 40, 0.08), 0px 4px 6px -2px rgba(16, 24, 40, 0.03)'
          }
          marginTop={gap}
          data-testid={__menuListTestId ?? 'hds.menu.list'}
          py={'0px'}
          fontSize={'14px'}
          color="neutrals.900"
        >
          {renderHeader}
          <Divider orientation="horizontal" />
          <MenuGroup>
            {!group &&
              (items as IMenuItemsSingle['items']).map(
                ({ command, label, icon }, idx) => {
                  return (
                    <>
                      <MenuItem command={command} icon={icon}>
                        {label}
                      </MenuItem>
                      {hasDivider && idx !== items.length - 1 && <Divider />}
                    </>
                  );
                }
              )}
            {group &&
              (items as IMenuItemsGroup['items']).map((item) => {
                return item.map(({ label, command, icon }, idx) => {
                  return (
                    <>
                      <MenuItem command={command} icon={icon}>
                        {label}
                      </MenuItem>
                      {idx === item.length - 1 && <Divider />}
                    </>
                  );
                });
              })}
          </MenuGroup>
        </MenuList>
      </Box>
    </Menu>
  );
};

export default Dropdown;
