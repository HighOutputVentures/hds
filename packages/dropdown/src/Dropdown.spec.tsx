import "@testing-library/jest-dom/extend-expect";

import { Box, ChakraProvider } from "@chakra-ui/react";
import { SettingIcon, UserIcon, ZapIcon } from "@highoutput/hds-icons";
import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import Dropdown from "./Dropdown";

const items = [
  { label: "View Profile", command: "⌘K->P", icon: <UserIcon /> },
  { label: "Settings", command: "⌘S", icon: <SettingIcon /> },
  { label: "Keyboard shortcuts", command: "?", icon: <ZapIcon /> },
];

describe("Dropdown", () => {
  beforeEach(() => {
    render(
      <ChakraProvider>
        <Dropdown items={items}>
          {({ onOpen }) => {
            return (
              <Box as="button" onClick={onOpen} data-testid="dropdown.button">
                Click
              </Box>
            );
          }}
        </Dropdown>
      </ChakraProvider>,
    );
  });

  it("Should render dropdown", async () => {
    await act(async () => {
      fireEvent.click(screen.getByTestId("dropdown.button"));
    });

    await waitFor(() => {
      expect(screen.getByTestId("hds.dropdown")).toBeInTheDocument();
    });
  });
});

describe("Snapshot", () => {
  const component = renderer.create(
    <ChakraProvider>
      <Dropdown items={items}>
        {({ onToggle }) => {
          return <Box as="button" onClick={onToggle} />;
        }}
      </Dropdown>
    </ChakraProvider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  component.unmount();
});
