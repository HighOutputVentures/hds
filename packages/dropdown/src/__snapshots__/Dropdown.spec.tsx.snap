// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <button
    className="css-0"
    onClick={[Function]}
  />,
  <span />,
  <span />,
]
`;
